src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -210,34 +201,62 @@ internal bool NullableSemanticAnalysisEnabled
         /// </summary>
         internal bool IsPeVerifyCompatEnabled => LanguageVersion < LanguageVersion.CSharp7_2 || Feature("peverify-compat") != null;
 
-        internal bool ShouldRunNullableWalker
+        /// <summary>
+        /// Returns true if nullable analysis is enabled in this compilation.
+        /// </summary>
+        /// <return>
+        /// Returns true if analysis is explicitly enabled for all methods;
+        /// false if analysis is explicitly disabled; and otherwise returns true
+        /// if there are any nullable-enabled contexts in the compilation.
+        /// </return>
+        internal bool IsNullableAnalysisEnabled
         {
             get
             {
-                if (!_lazyShouldRunNullableWalker.HasValue())
+                if (!_lazyShouldRunNullableAnalysis.HasValue())
+                {
+                    _lazyShouldRunNullableAnalysis = (GetNullableAnalysisValue() ?? hasEnabledContexts()).ToThreeState();
+                }
+                return _lazyShouldRunNullableAnalysis.Value();
+
+                bool hasEnabledContexts()
                 {
                     if (Options.NullableContextOptions != NullableContextOptions.Disable)
                     {
-                        _lazyShouldRunNullableWalker = ThreeState.True;
                         return true;
                     }
-
-                    foreach (var syntaxTree in SyntaxTrees)
-                    {
-                        if (((CSharpSyntaxTree)syntaxTree).HasNullableEnables())
-                        {
-                            _lazyShouldRunNullableWalker = ThreeState.True;
-                            return true;
-                        }
-                    }
-
-                    _lazyShouldRunNullableWalker = ThreeState.False;
+                    return SyntaxTrees.Any(tree => ((CSharpSyntaxTree)tree).HasNullableEnables());
```

Comment:
Does this method take into account projcet level enabling of nullability?

Reply:
The project-level setting is checked above with `if (Options.NullableContextOptions != NullableContextOptions.Disable) { return true; }`

---
In reply to: [541208903](https://github.com/dotnet/roslyn/pull/49876#discussion_r541208903) [](ancestors = 541208903)